### PR TITLE
Add baggage to B3 codec

### PR DESCRIPTION
**Pull request by pavolloffay**
Related to https://github.com/jaegertracing/jaeger/issues/755#issuecomment-393929718

* by default uses baggage- prefix
* it's enabled by default

Signed-off-by: Pavol Loffay <ploffay@redhat.com>

### CODE REVIEW

**Comment by pavolloffay** [comment]
we should decide whether baggage will be enabled by default or not. However our baggage management is different and separate from the codec. Which makes me think that we can remove `enableBaggage` from the codec and just use baggage manager. Note that baggage manager applies for all installed codecs. But doing this we are forced to by default enable baggage in B3. On the other hand it will be consistent with our codec.

Brave/Sleuth has to be configured explicitly which keys are whitelisted for propagation.

**Comment by yurishkuro** [comment]
+1 to integrate with baggage manager.

**Comment by pavolloffay** [comment]
Done in the last commit

**Comment by pavolloffay** [comment]
@yurishkuro could you please review?

File: jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java [hunk 1]
```diff
@@ -34,17 +37,34 @@
...
   protected static final String PARENT_SPAN_ID_NAME = "X-B3-ParentSpanId";
   protected static final String SAMPLED_NAME = "X-B3-Sampled";
   protected static final String FLAGS_NAME = "X-B3-Flags";
+  protected static final String BAGGAGE_NAME = "baggage-";
   // NOTE: uber's flags aren't the same as B3/Finagle ones
   protected static final byte SAMPLED_FLAG = 1;
   protected static final byte DEBUG_FLAG = 2;
 
+  private static final PrefixedKeys keys = new PrefixedKeys();
+  private final String baggageName;
```
Comment (black-adder) [inline jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java]: ditto call it prefix instead?

File: jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java [hunk 2]
```diff
@@ -80,12 +104,37 @@ public SpanContext extract(TextMap carrier) {
         if (entry.getValue().equals("1")) {
           flags |= DEBUG_FLAG;
         }
+      } else if (entry.getKey().startsWith(BAGGAGE_NAME)) {
```
Comment (black-adder) [inline jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java]: should this be `this.baggageName`?

File: jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java [hunk 3]
```diff
@@ -80,12 +104,37 @@ public SpanContext extract(TextMap carrier) {
...
+        if (baggage == null) {
+          baggage = new HashMap<String, String>();
+        }
+        baggage.put(keys.unprefixedKey(entry.getKey(), BAGGAGE_NAME), entry.getValue());
```
Comment (black-adder) [inline jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java]: ditto

**Comment by pavolloffay** [comment]
@black-adder PR updated.

File: jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java [hunk 4]
```diff
@@ -80,12 +104,37 @@ public SpanContext extract(TextMap carrier) {
...
+        }
+        baggage.put(keys.unprefixedKey(entry.getKey(), baggagePrefix), entry.getValue());
       }
     }
 
     if (null != traceId && null != parentId && null != spanId) {
-      return new SpanContext(traceId, spanId, parentId, flags);
+      SpanContext spanContext = new SpanContext(traceId, spanId, parentId, flags);
+      if (baggage != null) {
+        spanContext = spanContext.withBaggage(baggage);
```
Comment (black-adder) [inline jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java]: this is an interesting use case, what if only baggage is propagated without the rest of the trace context?

Comment (pavolloffay) [reply jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java]: At the moment it's an invalid span context - null

**Pull request merged**